### PR TITLE
Small speed up in IndicesQueryCache#getStats

### PR DESCRIPTION
**Pull request by DaveCTurner**
We don't need to build a map of all the shards just to compute the
shared RAM usage. With this commit we just compute the values for that
calculation directly.

This computation is still O(N²) in shard count (#97222) but with a much
smaller constant now.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/indices/IndicesQueryCache.java [hunk 1]
```diff
@@ -85,34 +85,42 @@ public IndicesQueryCache(Settings settings) {
...
-        for (Map.Entry<ShardId, Stats> entry : shardStats.entrySet()) {
-            stats.put(entry.getKey(), entry.getValue().toQueryCacheStats());
+    private static QueryCacheStats toQueryCacheStatsSafe(@Nullable Stats stats) {
+        return stats == null ? new QueryCacheStats() : stats.toQueryCacheStats();
+    }
+
+    private long getShareOfAdditionalRamBytesUsed(double cacheSize) {
+        // We also have some shared ram usage that we try to distribute proportionally to their number of cache entries of each shard.
+
+        if (shardStats.isEmpty()) {
```
Comment (original-brownbear) [inline server/src/main/java/org/elasticsearch/indices/IndicesQueryCache.java:95]: Is this worth it as an optimization? Seems like just an empty loop below should be good enough?

**Review by original-brownbear — APPROVED** [review]
LGTM! 

CI failure was me and should be fixed by merging in `main` sorry about that :)

File: server/src/main/java/org/elasticsearch/indices/IndicesQueryCache.java (hunk 1 above)
Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/indices/IndicesQueryCache.java]: Yeah fair point. I saw a couple of other fast-path possibilities tho, see 630f064.

**Comment by DaveCTurner** [comment]
[630f064](https://github.com/elastic/elasticsearch/pull/98900/commits/630f0640b6b16cb70ea7eace7cf2ce3ee8142d3c) changes it enough to be worth another quick look.

**Review by original-brownbear — APPROVED** [review]
LGTM :)

**Pull request merged**